pkg/kv/kvserver/replica_range_lease.go

```diff
@@ -1236,14 +1236,17 @@ func (r *Replica) redirectOnOrAcquireLeaseForRequest(
 					}
 
 					if pErr != nil {
-						switch tErr := pErr.GetDetail().(type) {
-						case *roachpb.AmbiguousResultError:
+						goErr := pErr.GoError()
+						switch {
+						case errors.HasType(goErr, (*roachpb.AmbiguousResultError)(nil)):
 							// This can happen if the RequestLease command we sent has been
 							// applied locally through a snapshot: the RequestLeaseRequest
 							// cannot be reproposed so we get this ambiguity.
 							// We'll just loop around.
 							return nil
-						case *roachpb.LeaseRejectedError:
+						case errors.HasType(goErr, (*roachpb.LeaseRejectedError)(nil)):
+							var tErr *roachpb.LeaseRejectedError
+							errors.As(goErr, &tErr)
```

Comment:
Yeah, it could be, but then we're writing a long `if {} else if {}` error chain. Somehow I prefer switch-based branching. I think because it indicates (though this is no longer true with `errors.HasType` but is morally still true) that the branches are non-overlapping (i.e not like `if i > 50 {} else if i > 25 {}` but more `if 0 < i && i < 50 {} else if i >= 50 && i < 100 {} else if ...`). There isn't a really strong argument either way here I think. Maybe a slight argument in favor of your suggestion that since the order of the cases now matters but folks will likely not realize this, the `if else` approach is better because it supports no such illusion? I'll leave as is because the code is already written but you've given me something to think about for next time around.

bors r=erikgrinaker